rpc: distinguish between vsize and sigop-adjusted mempool vsize

glozow opened the pull request:
`CTxMemPoolEntry::GetTxSize()` returns the maximum between vsize and "sigop-adjusted size" which is used by mempool validation and mining code as a heuristic to avoid 2DKnapsacking the block weight and sigop limits.

Sigop-adjusted vsize is returned as the "vsize" value of a transaction for RPCs retrieving mempool entry information (e.g. `getmempoolentry` and`getrawmempool`) and mempool acceptance (`testmempoolaccept` and `submitpackage`). The [documentation](https://github.com/bitcoin/bitcoin/blob/322ec63b01499c1ec52d3912ee382ebd59f2366b/src/rpc/mempool.cpp#L256) for these RPCs usually says "virtual transaction size as defined in BIP 141" without mentioning the sigop adjustment. At least 1 user has expressed confusion in a [tweet](https://twitter.com/mononautical/status/1646166180145577990?s=20).

I think we should at least mention something in the docs about the sigop-adjusted size. But imo it would be better to provide 2 vsize results: one that is sigop-adjusted and one that isn't.

Seeking feedback on what might be most clear/helpful for users.

This PR's approach is to change the result for "vsize" to not include sigop limits. The sigop-adjusted vsize is returned as a new result, "mempool_vsize".

glozow committed:
[rpc] add vsize_bip141 for non-sigop-adjusted vsize

Add a result for users who are expecting BIP141 vsize without sigop
adjustment.

mononaut commented:
Concept ACK

Sjors commented:
Concept ACK on clarifying and returning both variants.

It would useful to peruse a few other projects to see how they use `getmempoolentry` and `getrawmempool`, to determine if this is a breaking change.

luke-jr commented:
Concept ACK, but should be split into two commits.

luke-jr commented:
On further thought, it might make more sense to correct the description and maybe add a `vsize_bip141` field. Consensus only deals with `weight` anyway, so vsize doesn't really exist outside policy (and BIP141).

glozow committed:
[doc] mention sigop-adjusted size for vsize RPC result

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -132,6 +132,7 @@ static RPCHelpMan testmempoolaccept()
                     {RPCResult::Type::BOOL, "allowed", /*optional=*/true, "Whether this tx would be accepted to the mempool and pass client-specified maxfeerate. "
                                                        "If not present, the tx was not fully validated due to a failure in another tx in the list."},
                     {RPCResult::Type::NUM, "vsize", /*optional=*/true, "Virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted (only present when 'allowed' is true)"},
+                    {RPCResult::Type::NUM, "mempool_vsize", /*optional=*/true, "Sigop-adjusted virtual transaction size used by mempool and mining."},
```
Added mention of -bytespersigop

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -282,7 +285,8 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
 
-    info.pushKV("vsize", (int)e.GetTxSize());
+    info.pushKV("vsize", (int)GetVirtualTransactionSize(e.GetTx(), 0, 0));
+    info.pushKV("mempool_vsize", (int)e.GetTxSize());
```
removed

glozow commented:
> On further thought, it might make more sense to correct the description and maybe add a vsize_bip141 field.

Makes sense to me, changed. Also means no changes to the existing field apart from documentation.

> Concept ACK, but should be split into two commits.

Did this, one editing the doc and another to add the new field

Also rebased since it's been a few months.

luke-jr commented on src/rpc/mempool.cpp:
```diff
@@ -130,7 +130,9 @@ static RPCHelpMan testmempoolaccept()
                     {RPCResult::Type::BOOL, "allowed", /*optional=*/true, "Whether this tx would be accepted to the mempool and pass client-specified maxfeerate. "
                                                        "If not present, the tx was not fully validated due to a failure in another tx in the list."},
-                    {RPCResult::Type::NUM, "vsize", /*optional=*/true, "Virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted (only present when 'allowed' is true)"},
+                    {RPCResult::Type::NUM, "vsize", /*optional=*/true, "Maximum of sigop-adjusted size (-bytespersigop) and virtual transaction size as defined in BIP 141."},
```
I think we should just define "virtual size" to consider bytespersigop policy in general. (But the difficulty calculating it may be an issue)

glozow replied:
Are you suggesting we add a doc/policy/virtual_size.md?

ajtowns replied:
Perhaps doc/policy/feerates-and-vsize.md (or perhaps policy/mining-score.md) with the idea being to document how we prioritise txs for mining (namely via fee vs vsize), how vsize is calculated (weight scaled down by a factor of 4, then rounded up, and only sigop-adjusted when that is possible), and that for normal transactions sigop-adjustment doesn't actually change anything. Could also add that `-bytespersigop` allows tweaking how the sigop-adjustment works, and why that's important to do and can be a bad thing to change (bad fee estimation, suboptimal block templates).

stevenroose commented:
Concept ack.

Kinda unfortunate that the "vsize" field is still kinda broken IMO, only it's backwards compatible.

In fact, arguably since you changed the documentation, it is now also a breaking change in the API spec. And arguably the field not being implemented as it's documented could be seen as a bug and fixing the field to actually be what it says to be is a bugfix.

Personally I'd change the semantics of the vsize field to keep it "as per 141" and add something else like "vsize_sigops" for the alternative weird mempool-specific value.

stevenroose commented:
I suppose this is only relevant for pre-taproot (pre-segwit?) sigops, right? Because tapscript introduces this "validation weight/budget" concept? So I guess eventually the confusion will mostly disappear :)

kevkevinpal commented:
I took a look at some places where we could add tests for `vsize_bip141` and made this commit [4ab1f14](https://github.com/kevkevinpal/bitcoin/pull/3/commits/4ab1f1416bdc5830c1b7328d1ce81620e358db40) 

feel free to cherry commit or not use

ajtowns commented on src/rpc/mempool.cpp:
```diff
@@ -840,7 +845,8 @@ static RPCHelpMan submitpackage()
                         {RPCResult::Type::STR_HEX, "other-wtxid", /*optional=*/true, "The wtxid of a different transaction with the same txid but different witness found in the mempool. This means the submitted transaction was ignored."},
-                        {RPCResult::Type::NUM, "vsize", "Virtual transaction size as defined in BIP 141."},
+                        {RPCResult::Type::NUM, "vsize", "Maximum of sigop-adjusted size (-bytespersigop) and virtual transaction size as defined in BIP 141."},
+                        {RPCResult::Type::NUM, "vsize_bip141", "Virtual transaction size as defined in BIP 141."},
```
PR description needs updating from "mempool_vsize"

ajtowns commented on src/rpc/mempool.cpp:
```diff
@@ -252,7 +255,8 @@ static RPCHelpMan testmempoolaccept()
     return {
-        RPCResult{RPCResult::Type::NUM, "vsize", "virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted."},
+        RPCResult{RPCResult::Type::NUM, "vsize", "maximum of sigop-adjusted size (-bytespersigop) and virtual transaction size as defined in BIP 141."},
+        RPCResult{RPCResult::Type::NUM, "vsize_bip141", "virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted."},
```
I think rather than adding `vsize_bip141` to mempool RPCs, it would be better to (a) document mempool vsize as being sigop-adjusted, and (b) document rawtransaction vsize as **not** being sigop-adjusted? (Or, even better, do the sigop adjustment in cases where we can -- getrawtransaction of txs in blocks could pull up the utxo being spent via rev*.dat and calculate the sigop adjustment?)

Could be sensible to add "sigopsize" for the cases where we have a sigop count -- in that case vsize is simply max(floor((weight+3)/4), sigopsize), and the presence of a sigopsize with the exact same value as the weird vsize might be mostly self-explanatory as to what's going on. (And for the normal case where the weight dominates, it actually shows why the sigop adjustment isn't relevant)

glozow replied:
IIUC the suggestion is:
- Add docs for where vsize is sigop-adjusted and where it isn't.
- Add some doc/policy/?.md describing feerate, vsize, mining score
- Add an optional "sigopsize" result for `getrawtransaction` and the mempool RPCs when we're able to provide it

Will work on this

kristapsk commented:
Concept ACK

hebasto commented:
Concept ACK.

glozow commented:
Sorry, I kept thinking I would get back to this but haven't. Closing + marking up for grabs.
Note for anybody who wants to pick this up, note you'll want to write the doc described in https://github.com/bitcoin/bitcoin/pull/27591#discussion_r1378636536

fanquake commented:
This was picked up in #32800.
